Automattic/wp-calypso
# Signup: Use Redux for handling/persisting signup progress

Part of step 2 for #6709; builds on #25645. See p9xfpQ-8z-p2 for further discussion.

This change reworks the SignupProgressStore to use Redux instead of Flux. The SignupProgressStore is responsible for tracking and persisting the user's progression through our various signup flows.

Previously, this store persisted the current user's signup progress into localStorage via the [store.js](https://www.npmjs.com/package/store). It now relies on Calypso's [Redux state persistence mechanism](https://github.com/Automattic/wp-calypso/blob/master/client/state/initial-state.js#L166) which relies on [localforage](https://www.npmjs.com/package/localforage).

Existing signup flow behavior should remain identical as before.

Some notes: 

1. The signup flow is designed to cross the logged-out/logged-in boundary. Since the store persisted in logged-out state isn't loaded in the logged-in state, I'm assuming that the progress store is only useful during the signup process to the logged-out user. Once a user completes signup by completing the `user` step, we discard the logged-out state.

2. Upon construction of the signup flow controller, we invoke a function to discard steps not associated with the current flow. This ensures that we clear out stale/invalid steps persisted by the progress state in the Redux store.

# Testing instructions
1. Spin up this branch either on the [live branch](https://calypso.live/?branch=reduxify/signup-progress-store-part-2) or locally.
2. Go through the entirety of a signup flow such as [`/start`](http://calypso.localhost:3000/start/domains). 
3. Verify that you are correctly redirected following the signup flow. Depending on your A/B test configuration, this is either your new site or the new user checklist page.
